### ISSUE TITLE
Intermediate aggregation not being applied with union all

### ISSUE THREAD

**Issue by klaussfreire** [prestodb/presto#4831]
This is as simplified as I've been able to get it.

We've got two HDFS-backed tables, in parquet, hive catalog:

```
presto:evs> describe all_events_monthly;
       Column       |  Type   |    Comment    
--------------------+---------+---------------
...
 month_partition    | varchar | Partition Key 
 app_partition      | bigint  | Partition Key 
 ref_type_partition | varchar | Partition Key 
(17 rows)

presto:evs> describe all_install_hash_mapping_monthly;
       Column       |  Type   |    Comment    
--------------------+---------+---------------
...
 month_partition    | varchar | Partition Key 
 app_partition      | bigint  | Partition Key 
 ref_type_partition | varchar | Partition Key 
(11 rows)
```

(removed irrelevant columns)

This query ought to be performing preaggregation, but it isn't, it's transmitting the raw rows (which is far less efficient) to an aggregation remote step:

```
presto:evs> explain (type distributed) select approx_distinct(val) from (select '1' as val from all_events_monthly union all select '2' as val from all_install_hash_mapping_monthly) as "tab";
                                                                                                                                                                                                   Query Plan                                
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Fragment 0 [SINGLE]                                                                                                                                                                                                                         
     Output layout: [approx_distinct]                                                                                                                                                                                                        
     - Output[_col0] => [approx_distinct:bigint]                                                                                                                                                                                             
             _col0 := approx_distinct                                                                                                                                                                                                        
         - Aggregate => [approx_distinct:bigint]                                                                                                                                                                                             
                 approx_distinct := "approx_distinct"("expr_11")                                                                                                                                                                             
             - RemoteSource[1,2] => [expr_11:varchar]                                                                                                                                                                                        
                                                                                                                                                                                                                                             
 Fragment 1 [SOURCE]                                                                                                                                                                                                                         
     Output layout: [expr]                                                                                                                                                                                                                   
     - Project => [expr:varchar]                                                                                                                                                                                                             
             expr := '1'                                                                                                                                                                                                                     
         - TableScan[hive:hive:evs:all_events_monthly, originalConstraint = true] => []                                                                                                                                                    
                 LAYOUT: hive                                                                                                                                                                                                                
                 HiveColumnHandle{clientId=hive, name=app_partition, hiveType=int, hiveColumnIndex=-1, partitionKey=true}                                                                                                                    
                     :: [[1, 1592]]                                                                                                                                                                                                          
                 HiveColumnHandle{clientId=hive, name=month_partition, hiveType=string, hiveColumnIndex=-1, partitionKey=true}                                                                                                               
                     :: [[201512], [201601], [201602], [201603]]                                                                                                                                                                             
                 HiveColumnHandle{clientId=hive, name=ref_type_partition, hiveType=string, hiveColumnIndex=-1, partitionKey=true}                                                                                                            
                     :: [NULL, [android_id], [android_id_md5], [android_id_md5sha1], [android_id_sha1], [apple_ifa], [apple_ifa_md5], [apple_ifa_md5sha1], [apple_ifa_sha1], [google_advertising_id], [google_advertising_id_md5], [google_ad
                                                                                                                                                                                                                                             
 Fragment 2 [SOURCE]                                                                                                                                                                                                                         
     Output layout: [expr_9]                                                                                                                                                                                                                 
     - Project => [expr_9:varchar]                                                                                                                                                                                                           
             expr_9 := '2'                                                                                                                                                                                                                   
         - TableScan[hive:hive:evs:all_install_hash_mapping_monthly, originalConstraint = true] => []                                                                                                                                      
                 LAYOUT: hive                                                                                                                                                                                                                
                 HiveColumnHandle{clientId=hive, name=app_partition, hiveType=int, hiveColumnIndex=-1, partitionKey=true}                                                                                                                    
                     :: [[1, 1592]]                                                                                                                                                                                                          
                 HiveColumnHandle{clientId=hive, name=month_partition, hiveType=string, hiveColumnIndex=-1, partitionKey=true}                                                                                                               
                     :: [[201512], [201601], [201602], [201603]]                                                                                                                                                                             
                 HiveColumnHandle{clientId=hive, name=ref_type_partition, hiveType=string, hiveColumnIndex=-1, partitionKey=true}                                                                                                            
                     :: [NULL, [android_id], [android_id_md5], [android_id_md5sha1], [android_id_sha1], [apple_ifa], [apple_ifa_md5], [apple_ifa_md5sha1], [apple_ifa_sha1], [google_advertising_id], [google_advertising_id_md5], [google_ad
                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                             
(1 row)
```

This is Presto 0.136 running in AWS (EMR), it used to perform the preaggregation at least in 0.110, possibly also in later versions. (**edit**: no, it doesn't work with 0.110)

In fact, removing the union all (with a simpler query of course), it does perform the preaggregation:

```
presto:evs> explain (type distributed) select approx_distinct(val) from (select '2' as val from all_install_hash_mapping_monthly) as "tab";
                                                                                                                                                                                                     Query Plan                              
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Fragment 0 [SINGLE]                                                                                                                                                                                                                         
     Output layout: [approx_distinct]                                                                                                                                                                                                        
     - Output[_col0] => [approx_distinct:bigint]                                                                                                                                                                                             
             _col0 := approx_distinct                                                                                                                                                                                                        
         - Aggregate(FINAL) => [approx_distinct:bigint]                                                                                                                                                                                      
                 approx_distinct := "approx_distinct"("approx_distinct_6")                                                                                                                                                                   
             - RemoteSource[1] => [approx_distinct_6:HyperLogLog]                                                                                                                                                                            
                                                                                                                                                                                                                                             
 Fragment 1 [SOURCE]                                                                                                                                                                                                                         
     Output layout: [approx_distinct_6]                                                                                                                                                                                                      
     - Aggregate(PARTIAL) => [approx_distinct_6:HyperLogLog]                                                                                                                                                                                 
             approx_distinct_6 := "approx_distinct"("expr")                                                                                                                                                                                  
         - Project => [expr:varchar]                                                                                                                                                                                                         
                 expr := '2'                                                                                                                                                                                                                 
             - TableScan[hive:hive:evs:all_install_hash_mapping_monthly, originalConstraint = true] => []                                                                                                                                  
                     LAYOUT: hive                                                                                                                                                                                                            
                     HiveColumnHandle{clientId=hive, name=app_partition, hiveType=int, hiveColumnIndex=-1, partitionKey=true}                                                                                                                
                         :: [[1, 1592]]                                                                                                                                                                                                      
                     HiveColumnHandle{clientId=hive, name=month_partition, hiveType=string, hiveColumnIndex=-1, partitionKey=true}                                                                                                           
                         :: [[201512], [201601], [201602], [201603]]                                                                                                                                                                         
                     HiveColumnHandle{clientId=hive, name=ref_type_partition, hiveType=string, hiveColumnIndex=-1, partitionKey=true}                                                                                                        
                         :: [NULL, [android_id], [android_id_md5], [android_id_md5sha1], [android_id_sha1], [apple_ifa], [apple_ifa_md5], [apple_ifa_md5sha1], [apple_ifa_sha1], [google_advertising_id], [google_advertising_id_md5], [googl
                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                             
(1 row)
```

I haven't tested the latest versions, but I've checked the release notes and I haven't seen anything like this mentioned.

Seems like a regression. Applying the partial aggregation step should be perfectly safe with UNION ALL.

**Comment by klaussfreire** [comment]
Let me know if there is a way in which I can help.

**Comment by martint** [comment]
This could be related to b128446e1c03b0233322f7716e2534936858ff75.

**Comment by klaussfreire** [comment]
Doesn't seem to. We have `push_table_write_through_union` defaulted to true in our cluster:

```
presto:fault> show session;
                  Name                   |    Value    |   Default   |  Type   |                                   Description                                    
-----------------------------------------+-------------+-------------+---------+----------------------------------------------------------------------------------
 columnar_processing                     | false       | false       | boolean | Use columnar processing                                                          
 columnar_processing_dictionary          | false       | false       | boolean | Use columnar processing with optimizations for dictionaries                      
 dictionary_aggregation                  | false       | false       | boolean | Enable optimization for aggregations on dictionaries                             
 distributed_index_join                  | false       | false       | boolean | Distribute index joins on join keys instead of executing inline                  
 distributed_join                        | true        | true        | boolean | Use a distributed join instead of a broadcast join                               
 execution_policy                        | all-at-once | all-at-once | varchar | Policy used for scheduling query tasks                                           
 hash_partition_count                    | 8           | 8           | bigint  | Number of partitions for distributed joins and aggregations                      
 optimize_hash_generation                | true        | true        | boolean | Compute hash codes for distribution, joins, and aggregations early in query plan 
 prefer_streaming_operators              | false       | false       | boolean | Prefer source table layouts that produce streaming operators                     
 push_table_write_through_union          | true        | true        | boolean | Parallelize writes when using UNION ALL in queries that write data               
 query_max_run_time                      | 100.00d     | 100.00d     | varchar | Maximum run time of a query                                                      
 redistribute_writes                     | true        | true        | boolean | Force parallel distributed writes                                                
 resource_overcommit                     | false       | false       | boolean | Use resources which are not guaranteed to be available to the query              
 task_aggregation_concurrency            | 1           | 1           | bigint  | Experimental: Default number of local parallel aggregation jobs per worker       
 task_default_concurrency                | 1           | 1           | bigint  | Experimental: Default number of local parallel jobs per worker                   
 task_hash_build_concurrency             | 1           | 1           | bigint  | Experimental: Default number of local parallel hash build jobs per worker        
 task_intermediate_aggregation           | false       | false       | boolean | Experimental: add intermediate aggregation jobs per worker                       
 task_join_concurrency                   | 1           | 1           | bigint  | Experimental: Default number of local parallel join jobs per worker              
 task_share_index_loading                | false       | false       | boolean | Share index join lookups and caching within a task                               
 task_writer_count                       | 1           | 1           | bigint  | Default number of local parallel table writer jobs per worker                    
 hive.force_local_scheduling             | false       | false       | boolean | Only schedule splits on workers colocated with data node                         
 hive.orc_max_buffer_size                | 8MB         | 8MB         | varchar | ORC: Maximum size of a single read                                               
 hive.orc_max_merge_distance             | 1MB         | 1MB         | varchar | ORC: Maximum size of gap between two reads to merge into a single read           
 hive.orc_stream_buffer_size             | 8MB         | 8MB         | varchar | ORC: Size of buffer for streaming reads                                          
 hive.parquet_optimized_reader_enabled   | false       | false       | boolean | Experimental: Parquet: Enable optimized reader                                   
 hive.parquet_predicate_pushdown_enabled | false       | false       | boolean | Experimental: Parquet: Enable predicate pushdown for Parquet                     
(26 rows)
```

**Comment by martint** [comment]
Looking at the code, I'm not sure this ever worked. Can you post the plan you got in older versions?

**Comment by klaussfreire** [comment]
We don't have 0.110 running any more, but let me see if I can launch an older version, and check the simple query. If I have to import real data to get the old plan (does Presto use estimates about table sizes?) it may take a while.

**Comment by klaussfreire** [comment]
I cannot reproduce the old plan on 0.125 with empty schemas (not sure whether emptiness is a factor).

If memory serves, 0.110 managed to do it, but maybe my memory doesn't serve. Even if it never worked, it should be perfectly safe to do, and would be a big efficiency improvement - although it would go from regression to feature request ;) 

I'm trying to test with 0.119 now, if that doesn't reproduce the old plan, I will have to manually launch a 0.110 which will have to wait until next week :(

**Comment by klaussfreire** [comment]
Well, I finally launched 0.110 and it exhibits the same behavior, so maybe you were right and it never worked. But without the UNION ALL it does work (as per the explain in the OP). It may have been a coincidence that we started using UNION ALL at the same time we upgraded the cluster.

So, this isn't a regression itself, but there's certainly an oportunity for a substantial speedup here if it can be made to work across UNION ALL.
